espressostreamer/espresso_streamer.go

```diff
@@ -0,0 +1,254 @@
+	}
+	fullCalldata := append([]byte{}, method.ID...)
+	fullCalldata = append(fullCalldata, calldata...)
+
+	_, err = s.l1Reader.Client().CallContract(
+		ctx,
+		ethereum.CallMsg{
+			To:   &s.espressoTEEVerifierAddr,
+			Data: fullCalldata,
+		}, nil,
+	)
+	if err != nil {
+		return fmt.Errorf("call to the contract failed: %w", err)
+	}
+
+	return nil
+}
+
+/**
+* Create a queue of messages from the hotshot to be processed by the node
+* It will sort the messages by the message index
+* and store the messages in `messagesWithMetadata` queue
+ */
+func (s *EspressoStreamer) queueMessagesFromHotshot(ctx context.Context) error {
+
+	if s.nextHotshotBlockNum == 0 {
+		// We dont need to check majority here  because when we eventually go
+		// to fetch a block at a certain height,
+		// we will check that a quorum of nodes agree on the block at that height,
+		// which wouldn't be possible if we were somehow are given a height
+		// that wasn't finalized at all
+		latestBlock, err := s.espressoClient.FetchLatestBlockHeight(ctx)
+		if err != nil {
+			log.Warn("unable to fetch latest hotshot block", "err", err)
+			return err
+		}
+		log.Info("Started node at the latest hotshot block", "block number", latestBlock)
+		s.nextHotshotBlockNum = latestBlock
+	}
+
+	nextHotshotBlockNum := s.nextHotshotBlockNum
+	arbTxns, err := s.espressoClient.FetchTransactionsInBlock(ctx, nextHotshotBlockNum, s.namespace)
+	if err != nil {
+		log.Warn("failed to fetch the transactions", "err", err)
+		return err
+	}
+	if len(arbTxns.Transactions) == 0 {
+		log.Info("No transactions found in the hotshot block", "block number", nextHotshotBlockNum)
+		return nil
+	}
+
+	for _, tx := range arbTxns.Transactions {
+		// Parse hotshot payload
+		attestation, userDataHash, indices, messages, err := arbutil.ParseHotShotPayload(tx)
+		if err != nil {
+			log.Warn("failed to parse hotshot payload, will retry", "err", err)
+			return err
+		}
+		// if attestation verification fails, we should skip this message
+		// Parse the messages
+		if len(userDataHash) != 32 {
+			log.Warn("user data hash is not 32 bytes")
+			continue
+		}
+		userDataHashArr := [32]byte(userDataHash)
+		err = s.verifyAttestationQuote(ctx, attestation, userDataHashArr)
+		if err != nil {
+			log.Warn("failed to verify attestation quote", "err", err)
+			continue
+		}
+		for i, message := range messages {
+			var messageWithMetadata arbostypes.MessageWithMetadata
+			err = rlp.DecodeBytes(message, &messageWithMetadata)
+			if err != nil {
+				log.Warn("failed to decode message, will retry", "err", err)
+				// Instead of returnning an error, we should just skip this message
+				continue
+			}
+			if indices[i] < s.currentMessagePos {
+				log.Warn("message index is less than current message pos, skipping", "messageIndex", indices[i], "currentMessagePos", s.currentMessagePos)
+				continue
+			}
+			// Only append to the slice if the message is not a duplicate
+			if s.findMessageIndexByPos(indices[i]) == -1 {
```

Comment:
I think we're handling similar tasks in different places:

- When retrieving Espresso transactions, we check for duplicates here. However, this only examines the queue and does not account for consumed transactions.
- The queue is sorted after new messages are appended.
- When creating blocks, we remove messages with outdated numbers and always select the first one from the queue.
These processes are correct but somewhat unclear, especially since they run in different goroutines, adding complexity.

I suggest keeping a single function that retrieves the first message in the queue at a given position while also removing messages with smaller positions.

Reply:
@ImJeremyHe okay feel free to implement this function!